### PR TITLE
Add ext.menus pagination module

### PR DESCRIPTION
**Pull request by krittick**
### 
Adds a pagination module, building on the initial work done by @LilJess13 and others in #68

Works with prefixed and slash commands, and can be sent as a response to an interaction.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

File: discord/ext/menus/pagination.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+            self.paginator.current_page -= 1
+        elif self.button_type == "next":
+            self.paginator.current_page += 1
+        elif self.button_type == "last":
+            self.paginator.current_page = self.paginator.page_count
+        await self.paginator.goto_page(interaction=interaction, page_number=self.paginator.current_page)
+
+
+class Paginator(discord.ui.View):
+    """Creates a paginator for a message that is navigated with buttons.
```
Comment (RPSMain) [inline discord/ext/menus/pagination.py:32]: i personally don't like "navigated" used here maybe use "filled" or "implemented with buttons."

File: discord/ext/menus/pagination.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+        # We're done adding standard buttons, so we can now add any specified custom view items below them
+        # The bot developer should handle row assignments for their view before passing it to Paginator
+        if self.custom_view:
+            for item in self.custom_view.children:
+                self.add_item(item)
+
+    async def send(self, messageable: abc.Messageable, ephemeral: bool = False):
+        """Sends a message with the paginated items.
+        Parameters
```
Comment (RPSMain) [inline discord/ext/menus/pagination.py:206]: don't think this would be accepted by sphinx like that
```suggestion

        Parameters
```

File: discord/ext/menus/pagination.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        ------------
+        messageable: :class:`discord.abc.Messageable`
+            The messageable channel to send to.
+        ephemeral: :class:`bool`
+            Choose whether or not the message is ephemeral. Only works with slash commands.
```
Comment (RPSMain) [inline discord/ext/menus/pagination.py:211]:
```suggestion
            Choose whether the message is ephemeral or not. Only works with slash commands.
```

File: discord/ext/menus/pagination.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            else await messageable.send(
+                content=page if isinstance(page, str) else None,
+                embed=page if isinstance(page, discord.Embed) else None,
+                view=self,
+            )
+        )
+
+    async def respond(self, interaction: discord.Interaction, ephemeral: bool = False):
+        """Sends an interaction response or followup with the paginated items.
+        Parameters
```
Comment (RPSMain) [inline discord/ext/menus/pagination.py:243]:
```suggestion
        
        Parameters
```

File: setup.py [hunk 5]
```diff
@@ -31,7 +32,7 @@
         p = subprocess.Popen(["git", "rev-parse", "--short", "HEAD"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
         out, err = p.communicate()
         if out:
-            version += "+g" + out.decode("utf-8").strip()
+            version += f"+g{out.decode('utf-8').strip()}"
```
Comment (RPSMain) [inline setup.py:35]: why was this changed?

**Review by RPSMain — CHANGES_REQUESTED** [review]

File: setup.py (hunk 5 above)
Comment (krittick) [reply setup.py:35]: I think my IDE ran Flynt on the file by mistake, though it would still be good to move concatenated strings to f-strings where appropriate anyway (as a separate PR though - I'll revert this change here).

File: discord/ext/menus/pagination.py (hunk 1 above)
Comment (krittick) [reply discord/ext/menus/pagination.py:32]: This one is on my list to redo - still has the original verbiage from the old PR.

File: discord/ext/menus/__init__.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+"""
+discord.ext.menus
+~~~~~~~~~~~~~~~~~~~~~
+An extension module to provide useful menu options.
+
+:copyright: 2021 Pycord-Development
```
Comment (RPSMain) [inline discord/ext/menus/__init__.py:6]: i forgot to add present lmao
```suggestion
:copyright: 2021-present Pycord-Development
```

**Review by RPSMain — APPROVED** [review]

**Comment by krittick** [comment]
Added two more key bits of functionality:
- `self.message` now stores the message that gets sent with the paginator
- `on_timeout` event added which disables all buttons associated with the paginator view (including custom view buttons) if `disable_on_timeout` is set to True (which is the default). 
  - This results in a better user experience for views that time out instead of users receiving the "This interaction failed" error.

**Review by VincentRPS — APPROVED** [review]
looks good to me to be merged will have to test tho

File: examples/app_commands/slash_groups.py [hunk 7]
```diff
@@ -5,7 +5,7 @@
 # If you use commands.Bot, @bot.slash_command should be used for
 # slash commands. You can use @bot.slash_command with discord.Bot as well
 
-math = bot.command_group(
+math = bot.create_group(
     "math", "Commands related to mathematics."
```
Comment (VincentRPS) [inline examples/app_commands/slash_groups.py:9]: Why is this changed?

Comment (krittick) [reply examples/app_commands/slash_groups.py:9]: It was merged in from `master` as part of #576 and I'm just bad at keeping my commits clean for this branch

**Review by Lulalaby — APPROVED** [review]

**Review by ChickenDevs — APPROVED** [review]
